### PR TITLE
 Add a new `shape` field to `ColumnSchema`

### PR DESCRIPTION
**Pull request by karlhigley**
This creates a place to store shape information for all dimensions of the data across both array/tensor and dataframe formats. In contrast to the existing "value_count" property (which only records the value counts of the lists in list field, this attribute is intended to capture the size of _all_ dimensions of the data (the batch dimension, the list lengths, embedding sizes, etc.)

A few significant design elements:
* Defining a `Shape` dataclass allows us to make them immutable, which we wouldn't get if we sub-classed `tuple` like some other frameworks do. For compatibility with those other frameworks (and just general ease of use), the `Shape` constructor accepts a tuple of dimensions, so it should be relatively straightforward to (for example) create a Merlin shape from a Tensorflow shape.
* Shapes are stored in dtypes, since shape information is how we know a particular column is a list, and several frameworks we use (e.g. cuDF, Feast) have either a general `List` dtype or specific list dtypes like `int32_list` for each element type. The API has been designed to hide this from calling code as much as possible, but it does influence how shapes work to some extent.
* The `is_list` and `is_ragged` flags are now computed from the shape and can no longer be set by providing them as constructor args. We tried to find a way to maintain that, but it's not possible to use dataclass init vars with the same name as a defined property. The best workaround for this we could come up with was to allow `is_list` and `is_ragged` to be set from the `properties` dictionary, in which case we do our best to infer a shape from the values of these flags.

### CODE REVIEW

File: merlin/schema/schema.py [hunk 1]
```diff
@@ -240,12 +264,14 @@ def value_count(self) -> Optional[Domain]:
...
-            other.dtype, is_list=other.is_list, is_ragged=other.is_ragged
+        col_schema = (
+            self.with_name(other.name)
+            .with_dtype(other.dtype)
+            .with_tags(other.tags)
+            .with_properties(other.properties)
+            .with_properties({"is_list": other.is_list, "is_ragged": other.is_ragged})
         )
-        col_schema = col_schema.with_name(other.name)
+        # TODO: Figure out if and how to merge shapes here too
```
Comment (EvenOldridge) [inline merlin/schema/schema.py:274]: Is this deliberately left to future work or do we need a way to merge shapes still?

**Review by EvenOldridge — APPROVED** [review]
I like the level of detail into the error messaging and the comprehensive tests.  Assuming the merge of shapes is meant to be future work this looks good.

File: merlin/schema/schema.py (hunk 1 above)
Comment (karlhigley) [reply merlin/schema/schema.py:274]: Yeah, this should get filled in. The tricky part is figuring out what it means to merge shapes. 😅

Since you can have unbounded dimensions with `min=0, max=None`, one possible way to define it would be to take the shape from each column schema and:
- Check that the shapes have the same number of dimensions, if not raise an error
- For each dimension:
  - Check if the two versions of the dim are identical, if so use either one
  - Check if one is bounded and the other is unbounded, if so use the bounded version
  - Otherwise, we have two bounded versions of the dim that don't match, and we can either:
    - Choose whether the left or right side of the merge gets priority and overwrites the other
    - Raise an error

File: merlin/dtypes/base.py [hunk 2]
```diff
@@ -103,7 +109,7 @@ def to(self, mapping_name: str):
             ) from exc
 
         try:
-            return mapping.from_merlin(self)
+            return mapping.from_merlin(replace(self, shape=None))
```
Comment (karlhigley) [inline merlin/dtypes/base.py:112]: This should use the `.without_shape` method defined below

File: merlin/dtypes/shape.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            for i, dim in enumerate(self.dims):
+                if isinstance(dim, Dimension):
+                    new_dim = dim
+                elif isinstance(dim, tuple) and len(dim) == 2:
+                    new_dim = Dimension(dim[0], dim[1])
+                elif isinstance(dim, int):
+                    new_dim = Dimension(dim, dim)
+                else:
+                    raise ValueError(
+                        "Invalid shape tuple format: {self.dims}. Each dimension is expected to be "
```
Comment (karlhigley) [inline merlin/dtypes/shape.py:83]: This should be an f-string

File: merlin/dtypes/shape.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    def min(self) -> Tuple:
+        return tuple(dim.min for dim in self.dims)
+
+    @property
+    def max(self) -> Tuple:
+        return tuple(dim.max for dim in self.dims)
+
+    @property
+    def fixed(self) -> Tuple:
+        return tuple(dim.min if dim.is_fixed else None for dim in self.dims)
```
Comment (karlhigley) [inline merlin/dtypes/shape.py:115]: This method returns the shape with only the fixed dimensions included and the rest marked with `None` as a placeholder. I'm a bit worried that the distinction between `fixed` and `is_fixed` isn't clear, but I'm not sure what else to call them to make it clearer.

File: merlin/schema/schema.py [hunk 5]
```diff
@@ -82,37 +82,47 @@ def __post_init__(self):
...
+    def shape(self):
+        return self.dtype.shape
+
+    @property
+    def is_list(self):
+        return self.dtype.shape.is_list
+
+    @property
+    def is_ragged(self):
+        return self.dtype.shape.is_ragged
```
Comment (karlhigley) [inline merlin/schema/schema.py:125]: We might want to add `is_fixed`/`is_variable` here too, so that you don't have to reach through the dtype and the shape to access them (`col_schema.dtype.shape.is_variable`)

**Comment by bschifferer** [comment]
I reviewed the code. It seems good to me/makes sense - however, I haven't worked in that deep level of Merlin/core to confident approve it

File: merlin/dtypes/shape.py (hunk 4 above)
Comment (oliverholworthy) [reply merlin/dtypes/shape.py:115]: what's the use-case for wanting to return this tuple of fixed dims sizes?

File: tests/unit/dtypes/test_shape.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+
+def test_min_max_val_dimension():
+    dim = Dimension(2, 3)
+    assert dim.min == 2
+    assert dim.max == 3
+
+
+def test_fixed_min_with_unbounded_max():
+    dim = Dimension(2)
```
Comment (oliverholworthy) [inline tests/unit/dtypes/test_shape.py:37]: It may make our API more familiar to new users to make `Dimension(2)` correspond to a fixed dimension of size 2 (min=2, max=2) instead of a ragged dimension with min size of 2

Comment (oliverholworthy) [reply tests/unit/dtypes/test_shape.py:37]: I haven't seen any mentions of use-cases motivating the addition of the `min` in the dimension size other than it being there before in the value count property. I guess it's a more generic version of a Dimension in tensorflow for example that is either a fixed size or ragged with unknown min/max value.

Comment (oliverholworthy) [reply tests/unit/dtypes/test_shape.py:37]: We may need to give some consideration to what the min/max means in the ragged case. For example the ValueCount operator in NVTabular would result in a schema  with min and max set to the min/max observed in the data it was fit on. However, that doesn't necessarily imply that the model or serving system is limited to those min/max values

Comment (karlhigley) [reply tests/unit/dtypes/test_shape.py:37]: > It may make our API more familiar to new users to make Dimension(2) correspond to a fixed dimension of size 2 (min=2, max=2) instead of a ragged dimension with min size of 2

We considered this, but it made the code rather difficult to write, so we settled on providing syntactic sugar when creating a `ColumnSchema` with the `dims` property. We expect people to either use that path or the `Shape` constructor, which offers the same functionality, instead of interacting with the Dimension constructor directly. In either of those places, you can supply `2` for a dim, which means a dimension of fixed size like you're envisioning.

Comment (karlhigley) [reply tests/unit/dtypes/test_shape.py:37]: > I haven't seen any mentions of use-cases motivating the addition of the min in the dimension size other than it being there before in the value count property. I guess it's a more generic version of a Dimension in tensorflow for example that is either a fixed size or ragged with unknown min/max value.

We're not quite there yet, but we should be able to directly validate a schema against a `DictArray`/`TensorTable` and match up the shapes of the data with the requirements in the schema. Allowing a non-zero minimum lets us differentiate between list fields that are allowed to be empty and those that aren't. I'm not sure how useful that will end up being though.

Comment (karlhigley) [reply tests/unit/dtypes/test_shape.py:37]: > We may need to give some consideration to what the min/max means in the ragged case. For example the ValueCount operator in NVTabular would result in a schema with min and max set to the min/max observed in the data it was fit on. However, that doesn't necessarily imply that the model or serving system is limited to those min/max values

Yeah, this is a good point. What I'm envisioning is that each Merlin component updates the schema and stores the updated version alongside its artifacts, so when we save a model if the model is fine with a column having a minimum size of zero, it should update the schema to say so.

Something like this:
```
Parquet files -> NVT inferred schema -> NVT DAG -> NVT annotated schema -> Parquet files w/ annotated schema

Annotated schema -> dataloader transforms -> dataloader output schema -> model input layer

*model training happens*

Saved model w/ input and output schemas -> Systems op
```

File: merlin/schema/schema.py [hunk 7]
```diff
@@ -82,37 +82,47 @@ def __post_init__(self):
...
+
+        if new_shape:
+            object.__setattr__(self, "dtype", self.dtype.with_shape(new_shape))
+
+    @property
+    def shape(self):
+        return self.dtype.shape
+
+    @property
+    def is_list(self):
```
[the remark this answers is not in the document]
Comment (karlhigley) [reply merlin/schema/schema.py:120]: > I'm wondering if we can avoid the initial breaking change from this PR by keeping the is_list, is_ragged as properties that get set appropriately in the post_init and update methods based on the shape?

Yeah, we could try it.

File: merlin/dtypes/shape.py (hunk 4 above)
Comment (karlhigley) [reply merlin/dtypes/shape.py:115]: Let's remove it and see if we need to add it at some point. Seemed like a handy utility for handling things like `(64, (0, None), 128)` where you have a batch from the dataloader containing embedding lists of indeterminate length, but maybe I've overthought that part.

File: merlin/schema/schema.py (hunk 1 above)
Comment (karlhigley) [reply merlin/schema/schema.py]: For now, I'm just going to give the same priority to the shape from `other` as the rest of the attributes. We'll see if that causes problems and adjust if necessary.

**Comment by karlhigley** [comment]
@oliverholworthy I updated this PR to keep `is_list` and `is_ragged` as properties that are set in the `__post_init__` to match the shape, which really does clean things up quite a bit. I think it still ends up being a breaking change that requires coordination with the other repos though, because validating that the provided flags are consistent with the shape turns up places where that isn't currently true.

File: tests/unit/dtypes/test_shape.py (hunk 6 above)
Comment (oliverholworthy) [reply tests/unit/dtypes/test_shape.py:37]: > We considered this, but it made the code rather difficult to write, so we settled on providing syntactic sugar when creating a `ColumnSchema` with the `dims` property. [...] you can supply `2` for a dim, which means a dimension of fixed size like you're envisioning.

I'd expect to be able to proivde `None` too as part of this syntactic sugar. For example, `Shape((1, None, 10))` to create a shape with 2 fixed and one ragged dimension. This currently raises a ValueError.

Comment (oliverholworthy) [reply tests/unit/dtypes/test_shape.py:37]: Supporting `None` as a dimension to the Shape constructor would help with making it _relatively straightforward to (for example) create a Merlin shape from a Tensorflow shape._

**Comment by oliverholworthy** [comment]
> I think it still ends up being a breaking change that requires coordination with the other repos though, because validating that the provided flags are consistent with the shape turns up places where that isn't currently true.

Looks like we'll be setting a requirement for the shape to be specified if `is_ragged=False` is specified, the fixed dimension will be required to be passed. Would it be possible or helpful with the migration to loosen that restriction further and only check the shape if it's used explictly (either with the `dims` passed to the `ColumnSchema` constructor or `with_shape`)

**Comment by karlhigley** [comment]
> Would it be possible or helpful with the migration to loosen that restriction further and only check the shape if it's used explictly?

I shifted the validation in this general direction, so that only information that's explicitly provided gets validated. I'm not sure about lifting the restriction that some kind of shape info has to be provided (through the dtype's shape, the dims, or the value counts) when `is_ragged=False` because that will put us in a position where we sometimes don't have a shape. That may complicate the downstream changes as we try to migrate over. 🤔

**Comment by oliverholworthy** [comment]
> because that will put us in a position where we sometimes don't have a shape. That may complicate the downstream changes as we try to migrate over

I suppose we're already in the position where we sometimes don't have a shape, since there's nothing in here currently to enforce having a value count when `is_ragged=False`. Adding this requirement in this PR forces us to address that now. I support that choice if we're commited to making the changes to the downstream libs right now. If it's mostly a question of how we'll identify which parts of the code need to change, we could achieve that with an optional strict mode or something along those lines.

And closely related to this: to what extent will this change make to loading previously saved schemas (e.g. from an NVTabular workflow and saved schema file). Will we be able to load schemas saved with a prior version after merging this? We'd presumably prefer to be backwards compatible if possible, however if it's a necessary breaking change across the saved file format, then we might need some extra docs about this when we release the next version.

**Comment by karlhigley** [comment]
> Adding this requirement in this PR forces us to address that now. I support that choice if we're commited to making the changes to the downstream libs right now. If it's mostly a question of how we'll identify which parts of the code need to change, we could achieve that with an optional strict mode or something along those lines.

We're thinking along similar lines then; we've disabled the `is_ragged=False` without shape info validation for now, and will use it to identify the places where changes are needed in further dev work. 👍🏻 

> to what extent will this change make to loading previously saved schemas (e.g. from an NVTabular workflow and saved schema file). Will we be able to load schemas saved with a prior version after merging this? We'd presumably prefer to be backwards compatible if possible, however if it's a necessary breaking change across the saved file format, then we might need some extra docs about this when we release the next version.

I believe the change is backward but maybe not forward compatible, since the allowed values of value counts are slightly different in a more permissive direction after this change.

**Review by oliverholworthy — APPROVED** [review]
Looks like this is in good shape! ✨
